### PR TITLE
Proposal to enable more analyzer rules

### PR DESCRIPTION
**Pull request by bozturkMSFT**
This PR contains the rules I propose we enable. Feel free to comment on individual rules to discuss more. If you think it will be better, we can setup a meeting as well.

Some details:
- Almost all of the rules are about styling.
- Some can be auto-fixed in IDE for the whole solution, but not all
- Ordering of the type members requires a great deal of changes which cannot be automatically done by IDE. This maybe better fixed for the whole solution in a single PR.

This PR is for discussions only and not intended to go into master branch until template discovery PR is merged. So, there is no rush for this one.

### CODE REVIEW

File: eng/CodeAnalysis.ruleset [hunk 1]
```diff
@@ -24,9 +24,9 @@
     <Rule Id="CA1034" Action="None" />          <!-- Nested types should not be visible -->
     <Rule Id="CA1036" Action="None" />          <!-- Override methods on comparable types -->
     <Rule Id="CA1040" Action="None" />          <!-- Avoid empty interfaces -->
-    <Rule Id="CA1041" Action="None" />          <!-- Provide ObsoleteAttribute message -->
+    <Rule Id="CA1041" Action="Info" TemporarilyDisabled="true" />          <!-- Provide ObsoleteAttribute message -->
     <Rule Id="CA1043" Action="None" />          <!-- Use Integral Or String Argument For Indexers -->
-    <Rule Id="CA1044" Action="None" />          <!-- Properties should not be write only -->
+    <Rule Id="CA1044" Action="Info" TemporarilyDisabled="true" />          <!-- Properties should not be write only -->
```
Comment (bozturkMSFT) [inline eng/CodeAnalysis.ruleset:29]: This rule is already violated by `IRunnableProjectConfig.SourceFile` and requires updating the public interface. We may consider not enabling this rule, or disable it for this member only.

File: eng/CodeAnalysis.ruleset [hunk 2]
```diff
@@ -310,35 +310,35 @@
     <Rule Id="SA1141" Action="Warning" />       <!-- Use tuple syntax -->
     <Rule Id="SA1142" Action="Warning" />       <!-- Refer to tuple elements by name -->
     <Rule Id="SA1200" Action="None" />          <!-- Using directive should appear within a namespace declaration -->
-    <Rule Id="SA1201" Action="None" />          <!-- Elements should appear in the correct order -->
-    <Rule Id="SA1202" Action="None" />          <!-- Elements should be ordered by access -->
-    <Rule Id="SA1203" Action="None" />          <!-- Constants should appear before fields -->
-    <Rule Id="SA1204" Action="None" />          <!-- Static elements should appear before instance elements -->
+    <Rule Id="SA1201" Action="Info" TemporarilyDisabled="true" />          <!-- Elements should appear in the correct order -->
```
Comment (vlada-shubina) [inline eng/CodeAnalysis.ruleset:313]: Is this about alphabetic order? 
If so, I'm not sure about it for some cases - for examples for tests I sometimes prefer them to be logically ordered - happy path goes first, edge cases last.

**Review by vlada-shubina — COMMENTED** [review]
In general, enabled checks makes sense, however it moves us out-of-sync with runtime (original idea).

Is it possible to split the custom rules which are different from runtime to separate file? (which overrules runtime checks)

File: eng/CodeAnalysis.ruleset (hunk 2 above)
Comment (bozturkMSFT) [reply eng/CodeAnalysis.ruleset:313]: No, this is about how fields should come before properties and methods. More details here: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1201.md

**Review by DavidKarlas — APPROVED** [review]
I think this changes look sane... I'm also fine with changing API of IRunnableProjectConfig

**Comment by bozturkMSFT** [comment]
> In general, enabled checks makes sense, however it moves us out-of-sync with runtime (original idea).
> 
> Is it possible to split the custom rules which are different from runtime to separate file? (which overrules runtime checks)

Splitting the ruleset files don't work, unfortunately. Rules from the last specified file is used, the others are ignored.
I added xml attribute "TESpecific" to all the rules that differ from the ones in dotnet/runtime. This should help us identify the changes.

**Review by vlada-shubina — APPROVED** [review]

**Review by donJoseLuis — APPROVED** [review]

**Comment by bozturkMSFT** [comment]
I just noticed that these rules have the severity "info" anyway. So it should be OK to merge this, and update severity to "warning" as we refactor. Merging now.
